velox/functions/sparksql/RegexFunctions.cpp

```diff
@@ -16,37 +16,62 @@
 #include "velox/functions/lib/Re2Functions.h"
 
 #include "velox/common/base/Exceptions.h"
+#include "velox/expression/VectorWriters.h"
 
 namespace facebook::velox::functions::sparksql {
 namespace {
 
-// Blocks patterns that contain character class union, intersection, or
-// difference because these are not understood by RE2 and will be parsed as a
-// different pattern than in java.util.regex.
-void ensureRegexIsCompatible(
-    const char* functionName,
-    const VectorPtr& patternVector) {
-  if (!patternVector ||
-      patternVector->encoding() != VectorEncoding::Simple::CONSTANT) {
-    VELOX_USER_FAIL("{} requires a constant pattern.", functionName);
+using ::re2::RE2;
+
+template <typename T>
+re2::StringPiece toStringPiece(const T& s) {
+  return re2::StringPiece(s.data(), s.size());
+}
+
+// If v is a non-null constant vector, returns the constant value. Otherwise
+// returns nullopt.
+template <typename T>
+std::optional<T> getIfConstant(const BaseVector& v) {
+  if (v.encoding() == VectorEncoding::Simple::CONSTANT &&
+      v.isNullAt(0) == false) {
+    return v.as<ConstantVector<T>>()->valueAt(0);
   }
-  if (patternVector->isNullAt(0)) {
-    return;
+  return std::nullopt;
+}
+
+void checkForBadPattern(const RE2& re) {
+  if (UNLIKELY(!re.ok())) {
+    VELOX_USER_FAIL("invalid regular expression:{}", re.error());
   }
-  const StringView pattern =
-      patternVector->as<ConstantVector<StringView>>()->valueAt(0);
+}
+
+/**
+ * Validates the provided regex pattern to ensure its compatibility with the
+ * system. The function checks if the pattern uses features like character class
+ * union, intersection, or difference which are not supported.
+ *
+ * @param pattern The regex pattern string to validate.
+ * @param functionName (Optional) Name of the calling function to include in
+ * error messages. Default is "regex_replace".
```

Comment:
Do we need the default? Consider removing for readability. Is there any reason other functions do not validate the pattern?

Reply:
We should've used functionName instead of "Velox". Fixed in recent commit.

Right now, regex_replace is unique because it's the only spark regex function supporting non-constant patterns. So, to maintain regex compatibility, we introduced a special function that checks each pattern after they're decoded.

On the other hand, ensureRegexIsCompatible works on a VectorPtr, which is expected to have just one pattern. This function kicks in before we call RLIKE and REGEXP_EXTRACT. If we decide to allow non-constant patterns for them in the future, we could probably retire ensureRegexIsCompatible and just let them use checkForCompatiblePattern for every decoded pattern.